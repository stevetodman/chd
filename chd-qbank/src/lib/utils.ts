export function formatMs(ms?: number | null): string {
  if (ms === undefined || ms === null) return "â€“";
  const seconds = Math.max(0, Math.floor(ms / 1000));
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return mins > 0 ? `${mins}m ${secs}s` : `${secs}s`;
}

export function clampMs(ms: number): number {
  return Math.min(600_000, Math.max(0, ms));
}

export function percentage(part: number, total: number): string {
  if (total === 0) return "0%";
  return `${Math.round((part / total) * 100)}%`;
}

export function classNames(...classes: Array<string | false | null | undefined>) {
  return classes.filter(Boolean).join(" ");
}
