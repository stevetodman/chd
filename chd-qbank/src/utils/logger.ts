/**
 * Lightweight logging helper that standardizes console output.
 *
 * This wrapper can later be replaced with a structured logger
 * such as Pino or Sentry without requiring broad refactors.
 */
const logger = {
  info: (...args: unknown[]) => {
    console.info('[INFO]', ...args);
  },
  warn: (...args: unknown[]) => {
    console.warn('[WARN]', ...args);
  },
  error: (...args: unknown[]) => {
    console.error('[ERROR]', ...args);
  },
  debug: (...args: unknown[]) => {
    if (process.env.NODE_ENV === 'production') {
      return;
    }

    console.debug('[DEBUG]', ...args);
  },
} as const;

export default logger;
