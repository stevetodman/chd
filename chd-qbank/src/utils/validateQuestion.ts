import { Question, QuestionT } from '../schema/question.schema';

export function validateQuestion(q: unknown): QuestionT {
  const parsed = Question.safeParse(q);

  if (!parsed.success) {
    throw parsed.error;
  }

  return parsed.data;
}
