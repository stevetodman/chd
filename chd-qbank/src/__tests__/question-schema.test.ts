import { describe, it, expect } from "vitest";
import fs from "fs";
import path from "path";
import { validateQuestion } from "../utils/validateQuestion";

const ROOT = process.cwd();
const EXAMPLE_DIR = path.join(ROOT, "chd-qbank", "content", "questions");
const MEDIA_DIR = path.join(ROOT, "chd-qbank", "public", "media");

type RawQuestion = Record<string, unknown> & {
  choices: Array<Record<string, unknown>>;
  mediaBundle?: string[] | null;
  offlineRequired?: boolean;
  stem?: string;
  explanation?: string;
  references?: string[] | null;
};

function loadQuestions(): RawQuestion[] {
  if (!fs.existsSync(EXAMPLE_DIR)) return [];
  const walk = (dir: string): string[] =>
    fs.readdirSync(dir).flatMap((f) => {
      const p = path.join(dir, f);
      return fs.statSync(p).isDirectory() ? walk(p) : [p];
    });
  return walk(EXAMPLE_DIR)
    .filter((f) => f.endsWith(".json"))
    .map((f) => JSON.parse(fs.readFileSync(f, "utf8")) as RawQuestion);
}

describe("QBank question shape & assets", () => {
  const questions = loadQuestions();

  it("every question matches schema", () => {
    for (const q of questions) expect(() => validateQuestion(q)).not.toThrow();
  });

  it("stable top-level keys (snapshot)", () => {
    const shapes = questions.map((q) => Object.keys(q).sort());
    expect(shapes).toMatchSnapshot();
  });

  it("choices have expected keys", () => {
    for (const q of questions) {
      const keys = q.choices.flatMap((choice) => Object.keys(choice)).sort();
      expect(new Set(keys)).toEqual(new Set(["alt","id","isCorrect","label","mediaRef","text"]));
    }
  });

  it("media files exist when referenced", () => {
    for (const q of questions) {
      for (const file of q.mediaBundle ?? []) {
        const p = path.join(MEDIA_DIR, file);
        expect(fs.existsSync(p)).toBe(true);
      }
    }
  });

  it("offlineRequired items avoid remote-only refs", () => {
    for (const q of questions) {
      if (q.offlineRequired) {
        const all = [q.stem, q.explanation, ...(q.references ?? [])].join(" ");
        expect(all).not.toMatch(/https?:\/\//i);
      }
    }
  });
});
