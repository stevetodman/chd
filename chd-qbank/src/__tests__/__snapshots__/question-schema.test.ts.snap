// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QBank question shape & assets > stable top-level keys (snapshot) 1`] = `[]`;
